common/authorization/oauthAuthorizer.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package authorization
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/cristalhq/jwt/v3"
+	"go.uber.org/yarpc"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/config"
+)
+
+type oauthAuthority struct {
+	authorizationCfg config.OAuthAuthorizer
+}
+
+type jwtClaims struct {
+	Sub        string
+	Name       string
+	Permission string
+	Domain     string
+	Iat        int64
+}
+
+// NewOAuthAuthorizer creates a no-op authority
+func NewOAuthAuthorizer(
+	authorizationCfg config.OAuthAuthorizer,
+) Authorizer {
+	return &oauthAuthority{
+		authorizationCfg: authorizationCfg,
+	}
+}
+
+func (a *oauthAuthority) Authorize(
+	ctx context.Context,
+	attributes *Attributes,
+) (Result, error) {
+	call := yarpc.CallFromContext(ctx)
+	token := call.Header(common.AuthorizationTokenHeaderName)
+	// parseTokenAndVerify could either return the claims or a bool. I'm returning the claims currently
+	// because we might use it to populate values in the context
+	_, err := a.parseTokenAndVerify(token, attributes)
+	if err != nil {
+		return Result{Decision: DecisionDeny}, err
```

Comment:
returning error here is not a good idea. When error is not nil, AccessControlledWorkflowHandler will not look at the result. Even though the request will still be rejected, the error returned to user is InternalServiceError, and there will be a error log on the server side as "uncategoried error".

Check https://github.com/uber/cadence/blob/6f989a3d2089be82308b967d55213a1ab3f18296/service/frontend/accessControlledHandler.go#L773
and
https://github.com/uber/cadence/blob/6f989a3d2089be82308b967d55213a1ab3f18296/service/frontend/workflowHandler.go#L3725

IMO, some of the errors here should be deny result with no error:
error from `NewVerifierRS` and `NewVerifierRS` should be return as it.

The others should not be returned. Instead should return `Result{Decision: DecisionDeny}`. And I would suggest emit a Debug log for why the deny is happening, maybe:
```
logger.Debug("request is not authorized", tag.Error(err))
```
So that when this happen, users can enable debug log to see why if they need to

Reply:
I agree with your comment,

I refactored the code in 2 sections:

1. The one that validates the token (`parseToken()`). Technically speaking, this function would return only "panic" errors like:
`StringToRSAPublicKey`: If Public Key is incorrect
`NewVerifierRS`: If Algorithm is incorrect or if public key is not the right one
`ParseAndVerifyString`: If token is incorrect

All these functions as you suggested, are returning errors to the handler (therefore, the function is still returning an error)

2. The one that validates the claims (i.e expired token or domain mismatch).

This one on line 74 only returns `DecisionDeny` and adds the error to the log instead of returning it as you suggested.

Is this the expected approach?

Reply:
Almost, but not exactly yet. 

`ParseAndVerifyString` and `Unmarshal` are still users error so we shouldn't return error, but just deny with Debug log. 

Maybe you can separate `parseToken` into two parts: `prepareParsing` and `parseToken`.

Reply:
Done :) Let me know if you like more this approach

Reply:
LGTM